batch/src/main/java/is/hail/JVMEntryway.java

```diff
@@ -0,0 +1,222 @@
+    while (true) {
+      try (AFUNIXSocket socket = server.accept()) {
+        System.err.println("connection accepted");
+        DataInputStream in = new DataInputStream(socket.getInputStream());
+        DataOutputStream out = new DataOutputStream(socket.getOutputStream());
+        int nRealArgs = in.readInt();
+        System.err.println("reading " + nRealArgs + " arguments");
+        String[] realArgs = new String[nRealArgs];
+        for (int i = 0; i < nRealArgs; ++i) {
+          int length = in.readInt();
+          byte[] bytes = new byte[length];
+          System.err.println("reading " + i + ": length=" + length);
+          in.read(bytes);
+          realArgs[i] = new String(bytes);
+          System.err.println("reading " + i + ": " + realArgs[i]);
+        }
+
+        assert realArgs.length >= 2;
+        String classPath = realArgs[0];
+        String mainClass = realArgs[1];
+
+        String[] urlStrings = classPath.split(",");
+        ArrayList<URL> urls = new ArrayList<>();
+        for (int i = 0; i < urlStrings.length; ++i) {
+          File file = new File(urlStrings[i]);
+          urls.add(file.toURI().toURL());
+          if (file.isDirectory()) {
+            for (final File f : file.listFiles()) {
+              urls.add(f.toURI().toURL());
+            }
+          }
+        }
+        ClassLoader cl = new URLClassLoader(urls.toArray(new URL[0]));
+
+        System.err.println("have fresh classLoader for this job");
+        Class<?> klass = cl.loadClass(mainClass);
+        System.err.println("class loaded ");
+        Method main = klass.getDeclaredMethod("main", String[].class);
+        System.err.println("main method got ");
+
+        CompletionService<?> gather = new ExecutorCompletionService<Object>(executor);
+        Future<?> mainThread = null;
+        Future<?> shouldCancelThread = null;
+        Future<?> completedThread = null;
+        Throwable entrywayException = null;
+        try {
+          mainThread = gather.submit(new Runnable() {
+              public void run() {
+                ClassLoader oldClassLoader = Thread.currentThread().getContextClassLoader();
+                Thread.currentThread().setContextClassLoader(cl);
+                try {
+                  String[] mainArgs = new String[nRealArgs - 2];
+                  for (int i = 2; i < nRealArgs; ++i) {
+                    mainArgs[i-2] = realArgs[i];
+                  }
+                  main.invoke(null, (Object) mainArgs);
+                } catch (IllegalAccessException | InvocationTargetException e) {
+                  throw new RuntimeException(e);
+                } finally {
+                  Thread.currentThread().setContextClassLoader(oldClassLoader);
+                }
+              }
+            }, null);
+          shouldCancelThread = gather.submit(new Runnable() {
+              public void run() {
+                ClassLoader oldClassLoader = Thread.currentThread().getContextClassLoader();
```

Comment:
Can you explain the resetting of class loaders and why that's important? I assume it's a security reason. But you're making a new thread each time so why is this important?

Reply:
It is not a security reason. It's just that some old/poorly-written Java/Scala code will look at `Thread.currentThread().getContextClassLoader()` when it needs to load a class. I want that code to be able to see the Hail classes that we loaded.

It is not a new thread each time. The `gather` is an `ExecutorCompletionService` which is just a way to wait on multiple Runnables at once. That `ExecutorCompletionService` sends Runnables to the `executor`. The `executor` is a fixed size thread pool with exactly two threads. Since I always set the current thread's context class loader on startup, I don't really need to revert it (the next Runnable will always set it properly), but it seems prudent.